File: scripts/VerifyProvenance.cs
#:package Parquet.Net@6.0.3

using System;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text.Json;
using Parquet;

// -----------------------------------------------------------------------------
// VerifyProvenance.cs
// Verifies that all Git LFS Parquet files under benchmarks/data/ are:
// 1. Properly hydrated (not raw LFS text pointers < 1024 bytes).
// 2. Cryptographically identical to the recorded provenance.json SHA-256 digest.
// 3. Structurally valid with expected row count and column count.
// -----------------------------------------------------------------------------

string dataDir = FindBenchmarkDataDir();
string manifestPath = Path.Combine(dataDir, "provenance.json");

if (!System.IO.File.Exists(manifestPath))
{
    Console.Error.WriteLine($"❌ Provenance manifest not found: {manifestPath}");
    return 1;
}

using var manifestDoc = JsonDocument.Parse(System.IO.File.ReadAllText(manifestPath));
var root = manifestDoc.RootElement;

if (!root.TryGetProperty("datasets", out var datasetsElem) || datasetsElem.GetArrayLength() == 0)
{
    Console.Error.WriteLine($"❌ No datasets found in manifest: {manifestPath}");
    return 1;
}

int datasetCount = datasetsElem.GetArrayLength();
Console.WriteLine(
    $"🔍 Verifying provenance and cryptographic integrity for {datasetCount} dataset(s)..."
);

var errors = new List<string>();

foreach (var ds in datasetsElem.EnumerateArray())
{
    string id = ds.GetProperty("id").GetString()!;
    string filename = ds.GetProperty("filename").GetString()!;
    string filePath = Path.Combine(dataDir, filename);

    if (!System.IO.File.Exists(filePath))
    {
        errors.Add($"Missing file: {filePath}");
        continue;
    }

    var fileInfo = new FileInfo(filePath);
    if (fileInfo.Length < 1024)
    {
        errors.Add(
            $"{filename} is only {fileInfo.Length} bytes! Likely an unhydrated Git LFS pointer. "
                + $"Run 'git lfs pull' or verify 'lfs: true' in actions/checkout."
        );
        continue;
    }

    // Compute SHA-256
    string actualSha256;
    using (var fs = System.IO.File.OpenRead(filePath))
    {
        byte[] hashBytes = SHA256.HashData(fs);
        actualSha256 = Convert.ToHexString(hashBytes).ToLowerInvariant();
    }

    string expectedSha256 = ds.GetProperty("integrity").GetProperty("sha256").GetString()!;
    if (!string.Equals(actualSha256, expectedSha256, StringComparison.OrdinalIgnoreCase))
    {
        errors.Add(
            $"SHA-256 mismatch for {filename}: expected {expectedSha256}, got {actualSha256}"
        );
        continue;
    }

    // Check structural Parquet validity
    long expectedRows = ds.GetProperty("parquet_profile").GetProperty("row_count").GetInt64();
    int expectedCols = ds.GetProperty("parquet_profile").GetProperty("column_count").GetInt32();

    try
    {
        using var fs = System.IO.File.OpenRead(filePath);
        await using var reader = await ParquetReader.CreateAsync(fs);

        long actualRows = reader.Metadata?.NumRows ?? 0;
        int actualCols = reader.Schema.Fields.Count;

        if (actualRows != expectedRows)
        {
            errors.Add(
                $"Row count mismatch for {filename}: expected {expectedRows:N0}, got {actualRows:N0}"
            );
            continue;
        }

        if (actualCols != expectedCols)
        {
            errors.Add(
                $"Column count mismatch for {filename}: expected {expectedCols}, got {actualCols}"
            );
            continue;
        }

        Console.WriteLine(
            $"  ✅ {id} ({filename}): Verified SHA-256 & {actualRows:N0} rows, {actualCols} columns"
        );
    }
    catch (Exception ex)
    {
        errors.Add($"Failed to inspect Parquet metadata for {filename}: {ex.Message}");
    }
}

if (errors.Count > 0)
{
    Console.Error.WriteLine("\n❌ Provenance verification failed with the following error(s):");
    foreach (var err in errors)
    {
        Console.Error.WriteLine($"  - {err}");
    }
    return 1;
}

Console.WriteLine(
    "\n🎉 All benchmark datasets passed provenance and cryptographic integrity checks!"
);
return 0;

static string FindBenchmarkDataDir()
{
    string[] candidates =
    [
        Path.Combine(Directory.GetCurrentDirectory(), "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "benchmarks", "data"),
        Path.Combine(AppContext.BaseDirectory, "data"),
    ];

    foreach (var candidate in candidates)
    {
        string full = Path.GetFullPath(candidate);
        if (Directory.Exists(full) && System.IO.File.Exists(Path.Combine(full, "provenance.json")))
        {
            return full;
        }
    }

    var dir = new DirectoryInfo(Directory.GetCurrentDirectory());
    while (dir != null)
    {
        string candidate = Path.Combine(dir.FullName, "benchmarks", "data");
        if (
            Directory.Exists(candidate)
            && System.IO.File.Exists(Path.Combine(candidate, "provenance.json"))
        )
        {
            return Path.GetFullPath(candidate);
        }
        dir = dir.Parent;
    }

    return Path.GetFullPath(Path.Combine("benchmarks", "data"));
}

File: scripts/verify_provenance.py
# /// script
# dependencies = [
#     "pyarrow",
# ]
# ///

"""
Verifies that all Git LFS Parquet files under benchmarks/data/ are:
1. Properly hydrated (not raw LFS text pointers).
2. Cryptographically identical to the recorded provenance.json SHA-256 digest.
3. Structurally valid with expected row count and column count.
"""

import hashlib
import json
import os
import sys
import pyarrow.parquet as pq

BENCHMARK_DATA_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "benchmarks", "data"))
MANIFEST_PATH = os.path.join(BENCHMARK_DATA_DIR, "provenance.json")


def verify_provenance():
    if not os.path.exists(MANIFEST_PATH):
        print(f"❌ Provenance manifest not found: {MANIFEST_PATH}")
        sys.exit(1)

    with open(MANIFEST_PATH, "r", encoding="utf-8") as f:
        manifest = json.load(f)

    datasets = manifest.get("datasets", [])
    if not datasets:
        print(f"❌ No datasets found in manifest {MANIFEST_PATH}")
        sys.exit(1)

    print(f"🔍 Verifying provenance and integrity for {len(datasets)} dataset(s)...")
    errors = []

    for ds in datasets:
        file_name = ds["filename"]
        file_path = os.path.join(BENCHMARK_DATA_DIR, file_name)

        if not os.path.exists(file_path):
            errors.append(f"Missing file: {file_path}")
            continue

        file_size = os.path.getsize(file_path)
        if file_size < 1024:
            errors.append(
                f"{file_name} is only {file_size} bytes! Likely an unhydrated Git LFS pointer. "
                f"Run 'git lfs pull' or check actions/checkout lfs configuration."
            )
            continue

        # Check SHA-256
        hasher = hashlib.sha256()
        with open(file_path, "rb") as f:
            for chunk in iter(lambda: f.read(65536), b""):
                hasher.update(chunk)
        digest = hasher.hexdigest()

        expected_sha = ds["integrity"]["sha256"]
        if digest != expected_sha:
            errors.append(
                f"SHA-256 mismatch for {file_name}: "
                f"expected {expected_sha}, got {digest}"
            )
            continue

        # Check structural validity
        try:
            pf = pq.ParquetFile(file_path)
            actual_rows = pf.metadata.num_rows
            expected_rows = ds["parquet_profile"]["row_count"]
            if actual_rows != expected_rows:
                errors.append(
                    f"Row count mismatch for {file_name}: "
                    f"expected {expected_rows}, got {actual_rows}"
                )
                continue

            actual_cols = pf.metadata.num_columns
            expected_cols = ds["parquet_profile"]["column_count"]
            if actual_cols != expected_cols:
                errors.append(
                    f"Column count mismatch for {file_name}: "
                    f"expected {expected_cols}, got {actual_cols}"
                )
                continue

            print(f"  ✅ {ds['id']} ({file_name}): Verified SHA-256 & {actual_rows:,} rows")

        except Exception as ex:
            errors.append(f"Failed to read Parquet file {file_name}: {ex}")

    if errors:
        print("\n❌ Provenance verification failed with the following error(s):")
        for err in errors:
            print(f"  - {err}")
        sys.exit(1)

    print("\n🎉 All benchmark datasets passed provenance and cryptographic integrity checks!")


if __name__ == "__main__":
    verify_provenance()
